[ONNX] Fix export of copy_ operator

KsenijaS opened the pull request:
copy_operator before going into onnx exporter is being decomposed into aten::expand_as and aten::index_put.
There is a scenario where inputs to copy are not of the same type, but copy op in torch does implicit casting that is not currently reflected inside onnx exporter. This PR is adding casting inside index_put symbolic in case when tensor self is not of the same type as values.

KsenijaS committed:
fix copy export

jiafatom commented:
Please fix the build issue, thanks.

KsenijaS committed:
update helper file

KsenijaS committed:
add test

KsenijaS committed:
fix clang issue

neginraoof commented on torch/csrc/jit/passes/onnx/remove_inplace_ops_for_onnx.cpp:
```diff
@@ -656,10 +656,16 @@ void PrepareCopyForONNX(Node* node) {
         graph->insertNode(graph->createList(OptionalType::ofTensor(), {}))
             ->output();
 
+    auto* cast = graph->create(::c10::onnx::Cast, 1);
```
Please update the comments (line 652) to add this step.

BowenBao committed:
[ONNX] Redesign onnx pass to enable shape type dependent pattern conversion - cont (#51795)

With the introduction of ONNX shape inference, shape and type are inferred on the fly as operators get converted from ATen to ONNX when running symbolic function. This resolves the shape/type requirement for the symbolic functions. The pre-onnx passes however, can not be supported by shape inference, since at that stage the operators in the graph are still ATen operators.

This PR is to update the design of ONNX pass, to enable a mechanism of capturing subgraphs of ATen operators of certain patterns, and convert them later, when shape/type information of upstream operators are available.

The new design will require pre-onnx passes that need shape/type to be written in two parts, encapsulation and conversion.

    The encapsulation part will find the nodes of patterns, like how pre-onnx passes were written previously. But instead of converting the nodes, it will encapsulate them into a sub-block of a new placeholder node. This part is called before onnx pass, so it runs before calling symbolic functions.

    The conversion part will be called inside the onnx pass. In onnx pass, run_symbolic_func will be called for each node in topological order. When it reaches the placeholder node, the conversion part will be invoked. It will convert the nodes inside the sub-block based on pattern. By that time, it will have shape/type of upstream operators available. After the conversion is complete, the placeholder node will be removed, and nodes inside its sub-block converted. Run_symbolic_func will be called for these nodes, and they will be converted from ATen operator to ONNX operator.

This PR includes several other fixes, listed below.
* ~~replace helper.cpp with onnx_utils.cpp for holding utility functions.~~
* fix EraseNumberTypes on Bool type, the code was outdated that back then Bool type doesn't exist.
* ~~enable onnx shape inference in export with parameter/initializer data.~~
* other code clean ups.
* fix insertion of identity nodes for loop opset 13 sequence output.

~~PR depends on #51603~~

KsenijaS committed:
add comment

jiafatom committed:
[ONNX] Remove the last Cast in pow symbolic_opset9 (#52646)

Fixes #52436
For opset 9 of onnx Pow, if X is int32, Y is float, we will cast back to int32 which is consistent with X type.
However, pytorch is still float. The aten graph sometimes does not bind with the type for operators,
we are fine with the float type and don't want to cast back.
Even if X, Y are int32, the resulting float32 and int32 makes no difference.

neginraoof committed:
[ONNX] Support inplace operations on inplace indexing (#52063)

* [ONNX] Fix for sequence of mutations in blocks (#51577)

Fixes consecutive mutations in a tensor inside blocks.
Also, support append and pop in blocks.

* Support inplace operations + indexing

* Clean up old pass for remove mutations

* Add loop test

* Fixes for set attr in loops

* Removing the new jit API flag

* [ONNX] Redesign onnx pass to enable shape type dependent pattern conversion - cont (#51795)

With the introduction of ONNX shape inference, shape and type are inferred on the fly as operators get converted from ATen to ONNX when running symbolic function. This resolves the shape/type requirement for the symbolic functions. The pre-onnx passes however, can not be supported by shape inference, since at that stage the operators in the graph are still ATen operators.

This PR is to update the design of ONNX pass, to enable a mechanism of capturing subgraphs of ATen operators of certain patterns, and convert them later, when shape/type information of upstream operators are available.

The new design will require pre-onnx passes that need shape/type to be written in two parts, encapsulation and conversion.

    The encapsulation part will find the nodes of patterns, like how pre-onnx passes were written previously. But instead of converting the nodes, it will encapsulate them into a sub-block of a new placeholder node. This part is called before onnx pass, so it runs before calling symbolic functions.

    The conversion part will be called inside the onnx pass. In onnx pass, run_symbolic_func will be called for each node in topological order. When it reaches the placeholder node, the conversion part will be invoked. It will convert the nodes inside the sub-block based on pattern. By that time, it will have shape/type of upstream operators available. After the conversion is complete, the placeholder node will be removed, and nodes inside its sub-block converted. Run_symbolic_func will be called for these nodes, and they will be converted from ATen operator to ONNX operator.

This PR includes several other fixes, listed below.
* ~~replace helper.cpp with onnx_utils.cpp for holding utility functions.~~
* fix EraseNumberTypes on Bool type, the code was outdated that back then Bool type doesn't exist.
* ~~enable onnx shape inference in export with parameter/initializer data.~~
* other code clean ups.
* fix insertion of identity nodes for loop opset 13 sequence output.

~~PR depends on #51603~~

* Fix after merge

* clang

* Fix clang

* Fix clang

* Fix warning message.

* Fixes for non-model param attributes

* Fix for caffe2

* Additional test

* clang

* Skip test for lower opsets

* fix clang-tidy

* Update init.cpp

* Update remove_inplace_ops_for_onnx.cpp

* Update remove_inplace_ops_for_onnx.cpp

* Update remove_inplace_ops_for_onnx.cpp

* Fix for clang formatting

Co-authored-by: Bowen Bao <bowbao@microsoft.com>

jiafatom committed:
[ONNX] Symbolic shape inference (#51481)

This PR did symbolic shape inference, in the onnx pass _jit_pass_onnx_graph_shape_type_inference.
It creates a singleton ConstantValueMap.
It leverages constant folding technique and did a per-op based handling for ConstantValueMap.
As a byproduct, it enables fold_if pass for dynamic axes cases, typically for faster-rcnn etc.

The core change is in `torch/csrc/jit/passes/onnx/shape_type_inference.cpp` and `torch/csrc/jit/passes/onnx/constant_map.cpp`.

We usually need copy tensor to store in the ConstantValueMap, otherwise the underlying value may change. I see this issue in (1) from_blob (2) get value from Constant node.

fatcat-z committed:
Update test cases generated by make_test() method to support running them in script mode. (#52748)

* Update tests for test_gru_* at this moment.

* Update flake8 error.

* Update tests for test_gru_* at this moment.

* Update flake8 error.

* Update test_gru_* test cases only.

* Fix flake8 issue.

* Fix flake8 issue on test.

* Still disable test cases created by make_test.

* Update code to fix issue 'AttributeError: 'RecursiveScriptModule' object has no attribute 'forward'' for test_elman_* test cases.

* Add script model support for test_lstm_* test cases.

hwangdeyu committed:
[ONNX] enable several script unit tests using new jit passes (#51722)

* enable scripting related unit test

* fix flaske

* enable more tests

* fix interpolate and BCElogits test

* fix interpolate test ci

* add test_interpolate_upsample opset comment

* add interpolate_upsample tracing support below opset 9

Co-authored-by: hwangdeyu <deyhuang@qq.com>

neginraoof commented on torch/csrc/jit/passes/onnx/helper.cpp:
```diff
@@ -85,6 +85,39 @@ c10::optional<at::ScalarType> ONNXTypeToATenType(int32_t onnx_type) {
   return c10::optional<at::ScalarType>{};
 }
 
+int32_t ATenTypeToONNXType(c10::optional<at::ScalarType> aten_type) {
```
I think this function is already added to helper.cpp. Please rebase.

KsenijaS committed:
Merge branch 'onnx_ms_1' of https://github.com/pytorch/pytorch into fix_copy

KsenijaS committed:
update passes

BowenBao commented on torch/csrc/jit/passes/onnx/remove_inplace_ops_for_onnx.cpp:
```diff
@@ -340,20 +340,27 @@ void PrepareCopyForONNX(Node* node) {
 
+    auto* cast = graph->create(::c10::onnx::Cast, 1);
+    cast->addInput(node->input(1));
+    auto aten_type = node->input(0)->type()->cast<TensorType>()->scalarType();
```
need to add check for 1. if `cast<TensorType>()` returns nullptr, and 2. if `scalarType()` returns c10::nullopt.

BowenBao commented on torch/csrc/jit/passes/onnx/helper.h:
```diff
@@ -44,7 +44,7 @@ TORCH_API Value* addInputToBlock(Block* block);
 
 TORCH_API ONNX_NAMESPACE::TensorProto_DataType ATenTypeToOnnxType(
-    at::ScalarType at_type);
+    c10::optional<at::ScalarType> at_type);
```
after adding suggested check, this change can be reverted.

BowenBao committed:
Fix copy_ export (#53046)

Fixes export of torch.copy_

KsenijaS committed:
update passes

KsenijaS committed:
clang format

BowenBao committed:
[ONNX] Update assign output shape for nested structure and dict output (#52893)

Fixes dict output & nested tuple.

KsenijaS committed:
Merge branch 'onnx_ms_1' of https://github.com/pytorch/pytorch into fix_copy

shubhambhokare1 committed:
[ONNX] Add repeat_interleave symbolic (#52855)

- Add support for aten::repeat_interleave
- NOTE: Also adds fix for cases with split op where input tensor sizes are not known but _outputs is provided

BowenBao reviewed (approved):
LGTM, thanks

BowenBao commented:
Can you take a look at the test failures for ort_test2? Maybe due to new updates in base branch.

KsenijaS committed:
Merge branch 'onnx_ms_1' of https://github.com/pytorch/pytorch into fix_copy

KsenijaS committed:
update pass

KsenijaS committed:
[ONNX] Update ONNX documentation (#51362)

Add information about .data field

KsenijaS committed:
update pass

KsenijaS committed:
Merge branch 'onnx_ms_1' of https://github.com/pytorch/pytorch into fix_copy

KsenijaS committed:
fix type_as symbolic

KsenijaS committed:
fix flake8 failure

BowenBao committed:
[ONNX] Support torch.isinf, torch.any and torch.all export to ONNX (#53328)

Supported for ONNX export after opset 10.
This is not exportable to opsets < 10 due to
1. onnx::IsInf is introduced in opset 10
2. onnx::Equal does not accept float tensor prior to opset 11

neginraoof committed:
[ONNX] fix export of embedding with padding_idx (#53053)

fix export of embedding with padding_idx

shubhambhokare1 committed:
[ONNX] Update inputs/input_names formatting to avoid ValueError with scriptMethods (#53519)

fixes issue faced in #53506

BowenBao commented on torch/onnx/symbolic_opset9.py:
```diff
@@ -1479,19 +1479,17 @@ def index_copy(g, self, dim, index, source):
-        else:
-            raise RuntimeError('Unsupported: ONNX export of type_as for tensor '
-                               'of unknown dtype.')
+    if other_dtype is None:
```
This doesn't feel right, we cannot assume `Float` when type is not known. Do you see a case failing without doing this?

KsenijaS committed:
Merge branch 'onnx_ms_1' of https://github.com/pytorch/pytorch into fix_copy

KsenijaS committed:
revert type_as

BowenBao commented on torch/onnx/symbolic_opset11.py:
```diff
@@ -143,9 +143,14 @@ def index_put(g, self, indices_list_value, values, accumulate=False):
         dtype = sym_help.scalar_type_to_onnx.index(sym_help.cast_pytorch_to_onnx[dtype])
         dtype = sym_help.scalar_type_to_pytorch_type[dtype]
         zeros = g.op("ConstantOfShape", g.op("Shape", self), value_t=torch.tensor([0], dtype=dtype))
```
it appears you will need to cast self to dtype of values for both cases, if they don't match.

For zeros, it is constructed from ConstantOfShape. So instead of casting later, you can construct it with the decided dtype.

KsenijaS committed:
fix merge conflkict

KsenijaS committed:
fix index_put symbolic

KsenijaS committed:
fix merge

KsenijaS committed:
fix merge

KsenijaS committed:
fix merge

BowenBao commented:
please check onnx ci failure

KsenijaS committed:
update opset 11

KsenijaS committed:
update symbolic
